src/Docker/Runner/DataLoader/WorkspaceProviderFactory.php

```diff
@@ -35,97 +40,100 @@ public function getWorkspaceStaging(
         array $backendConfig,
         ?bool $useReadonlyRole,
         ?ExternallyManagedWorkspaceCredentials $externallyManagedWorkspaceCredentials,
-    ): AbstractWorkspaceProvider {
+    ): WorkspaceProviderInterface {
+        if (!in_array($stagingStorage, AbstractStrategyFactory::WORKSPACE_TYPES, true)) {
+            return new InvalidWorkspaceProvider($stagingStorage);
+        }
+
         if ($externallyManagedWorkspaceCredentials) {
             // Externally managed workspaces are persistent
-            $workspaceStaging = new ExistingWorkspaceStagingProvider(
+            $workspaceStaging = new ExistingWorkspaceProvider(
                 $this->workspacesApiClient,
                 $externallyManagedWorkspaceCredentials->id,
-                $externallyManagedWorkspaceCredentials->getDatabaseCredentials(),
+                new ExistingCredentialsProvider(
+                    $externallyManagedWorkspaceCredentials->getWorkspaceCredentials(),
+                ),
             );
             $this->logger->notice(sprintf(
                 'Using provided workspace "%s".',
                 $externallyManagedWorkspaceCredentials->id,
             ));
-        } elseif ($configId && ($stagingStorage === AbstractStrategyFactory::WORKSPACE_ABS)) {
-            // ABS workspaces are persistent, but only if configId is present
-            $workspaceStaging = $this->getPersistentWorkspace(
-                $component,
-                $configId,
-                'abs',
-                ABSWorkspaceCredentials::class,
-            );
-        } elseif ($configId && ($stagingStorage === AbstractStrategyFactory::WORKSPACE_REDSHIFT)) {
-            // Redshift workspaces are persistent, but only if configId is present
-            $workspaceStaging = $this->getPersistentWorkspace(
+            return $workspaceStaging;
+        }
+
+        $workspaceLoginType = null;
+        if ($stagingStorage === AbstractStrategyFactory::WORKSPACE_SNOWFLAKE && $component->useSnowflakeKeyPairAuth()) {
+            $workspaceLoginType = WorkspaceLoginType::SNOWFLAKE_SERVICE_KEYPAIR;
+        }
+
+        $workspaceBackendConfig = new WorkspaceBackendConfig(
+            $stagingStorage,
+            $backendConfig['type'] ?? null,
+            $useReadonlyRole,
+            NetworkPolicy::SYSTEM,
+            $workspaceLoginType,
+        );
+
+        // ABS & Redshift workspaces are persistent, but only if configId is present
+        // Persistent BigQuery workspaces are possible, but do not work well on connection side (shared buckets,
+        // read-only role). If fixed, uncomment this + add changes to DataLoader class and it will start working.
+        if ($configId && in_array($stagingStorage, [
```

Comment:
Jak jsem to postupne prehazoval na predelany `staging-provider`, dosel jsem k tomu, ze ty persistent workspaces vlastne vsehcny funguji stejne a jde je resit univerzalne.

Jediny, v cem se fakticky lisily, byly classes, ktery se pouzivaly na predani credentials (`ABSWorkspaceCredentials` apod.). Ty se ale stejne jen naparsovaly z pole a hned zase vyplovly do pole, takze jsem je v ramci zjednoduseni uplne vynechal. Jediny downside, kteyr vidim, je, ze pokud by credentials byly uplne blbe (poslalo se napr. `password` misto `connectionString`), failnulo yb to az pri pokusu o prihlaseni (za behu jobu misto tady). Stejne jak kdyz jsou credentials neplatny (spatny `connectionString`).

Reply:
Plus to vypada, ze `WORKSPACE_ABS` s ebude deprekovat, takze za chvili tu stejne asi zbude jen Redshift.

Reply:
a hele no, tak nerisknem to a nevyhodíme ten redshift úplně? https://keboolaglobal.slack.com/archives/C054VS93U15/p1747039105720949?thread_ts=1747038017.498009&cid=C054VS93U15 nebo separé PR?

Reply:
pujde pryc, ale uz bych to nechal separatne.. tim, ze tu pak nebud7e zadny persistent backend, musime se domluvit, co obecne s podporou persistent workspaces. Jestli to ted uplne vyhodime z kodu a nekdy v budoucnu pripadne vytahnem z Git history a adaptujem na novou code base.. nebo to tam nechame, nejak nafejkovany, aby se to dalo dal z testu provolat.